### ISSUE TITLE
[Enhancement] Add feature "Play with Kodi"

### ISSUE THREAD

**Issue by ChrisPHL** [cemrich/zapp#108]
Hi there,

the feature "Play with Kodi" is known from NewPipe (YouTube App: https://github.com/TeamNewPipe/NewPipe). I would much appreciate to have that feature from the "Mediathek" tab since the typically Mediatheken Kodi Add-On's search functions do not work well with the TV's remote control.
Would you consider to add that feature?

Kind regards
Christian

**Comment by cemrich** [comment]
This is a cool feature that I will include in the future. Since then you can long press on a mediathek item and select "Teilen" - the Kodi app will pop up if it is installed. The cast icon however is a much nicer user interface :-)

**Comment by rugk** [comment]
When this thing replaces the usual Android share menu, I really don't like that. It may look nice first, but then you miss this new Android feature, where it displays the contacts to share to directly at the start of the share page.

**Comment by cemrich** [comment]
There seem to be some changes with Android Q: https://developer.android.com/preview/features/sharing I'll wait for them to settle.

**Comment by Bond246** [comment]
I really would like to add something to this topic.

Actualy every mediathek item has a sharing button in the top right corner. With this icon it's possible to share the video-link with every service that android supports... 
But sometimes (zdf for example) this top right sharing icon has not the highest quality. So you have to use one of the quality items in the bottom list. The sharing button of this individual qualitys has not the whole sharing app library. 

So to share the maximum quality with the kodi app i mostly have to go back to the web-version of mediathekview, search for the video and share the video.

It would be nicer if the individual quality sharing buttons have the same sharing library list like the top right corner sharing icon.

**Comment by cemrich** [comment]
> I really would like to add something to this topic.
> 
> Actualy every mediathek item has a sharing button in the top right corner. With this icon it's possible to share the video-link with every service that android supports...
> But sometimes (zdf for example) this top right sharing icon has not the highest quality. So you have to use one of the quality items in the bottom list. The sharing button of this individual qualitys has not the whole sharing app library.
> 
> So to share the maximum quality with the kodi app i mostly have to go back to the web-version of mediathekview, search for the video and share the video.
> 
> It would be nicer if the individual quality sharing buttons have the same sharing library list like the top right corner sharing icon.

@Bond246 You are right. The top corner button shares the medium quality url (because we can be sure it exists) as text/plain, whereas the quality buttons share as video/*. Technically the quality buttons work correct and they play nice with VLC, Yatse and the google video player. Sharing as plain text shows more apps, but most likely apps you don't want to share with.

Are you using Kore to share to Kodi? For some reason Kore does not react to video/* share intents: https://github.com/xbmc/Kore/blob/master/app/src/main/AndroidManifest.xml

Any thoughts on this implementation?
- Should the share button always share the maximum quality available?
- Should more apps show up for the quality share buttons?

**Comment by Bond246** [comment]
Hello,

you are right, i use kore which was more handy than yatse last time i was testing both of this apps.
So you think it might be a good idea to contact the kore devs to ask for video/* implementation?

From my point of view having always the best quality in top right corner sharing button would be the best idea.
Or maybe it would be a good idea to make this a user configuration?
I mean... i know that not all videos have the highes quality so i understand that you say it is a good idea to have only the smaller medium quality there...

And i think sometimes i would be good to copy just the quality url for sharing it via whatsapp, notes app and so on... So having more than only video apps is more useful.
Most of the android system sort the sharing menu by recent used apps.

**Comment by cemrich** [comment]
> you are right, i use kore which was more handy than yatse last time i was testing both of this apps.
> So you think it might be a good idea to contact the kore devs to ask for video/* implementation?

If kore can process video/* urls it should be able to accept this mime type. Maybe there is a good reason for this behavior but it doesn't hurt to ask.

> From my point of view having always the best quality in top right corner sharing button would be the best idea.

Yes, I'll do that. Best quality and medium quality for fallback.

> Or maybe it would be a good idea to make this a user configuration?

Nope. Most users won't have this problem. For all other users every additional option does make the ui more complicated.

> And i think sometimes it would be good to copy just the quality url for sharing it via whatsapp, 
notes app and so on... So having more apps than only video apps is more useful.
In this case you can use the top right button. There should be no need to share a low quality url to anything other than a video player of some kind.

> Most of the android system sort the sharing menu by recent used apps.

If this was the case I would not even discuss this issue. Zapp could happily share with all the apps. But on modern Android versions the sharing menu is a ever changing, inconsistent mess. Maybe Android Q will fix some of these issues.

**Comment by WebworkrNet** [comment]
Wouldn't it be possible to select the quality level beforehand, as is the case under "Download"? Then the handling would be uniform.

It is also currently unclear what the quality is when the share menu is used. A corresponding note is missing.

**Comment by WebworkrNet** [comment]
> Wouldn't it be possible to select the quality level beforehand, as is the case under "Download"? Then the handling would be uniform.
> 
> It is also currently unclear what the quality is when the share menu is used. A corresponding note is missing.

![20210530_100755](https://user-images.githubusercontent.com/43509007/120097171-5e9e7a00-c12f-11eb-98e9-fb7558ae0480.jpg)

**Comment by kernal64** [comment]
> There is a share button directly beneath the download button. There you can select the quality you want to share. I try to leave the main actions as easy as possible and want to avoid additional clicks for the main share button in the actionbar.

I assume, that _the main share button in the actionbar_ is the one in the upper right corner? The share button beneath the download button doesn't offer all relevant apps. To name a few, I'm missing "pyLoad", "FairEmail", "WhatsApp", "Nextcloud" and "Total Commander" in the list of receiving apps over there. They are available, when I use the main share button in the actionbar, but then medium resolution is selected, even if high and low are available.

IMHO, either all the same apps available with the icon in the upper right corner should be available with the share button directly beneath the download button or the quality selection dialog should show when using the the share icon in the upper right corner.

I understand, that you want to avoid additional clicks for the main share button in the actionbar. Maybe a "Share Quality"-option in the settings could be a good balance?

**Comment by cemrich** [comment]
Ah, sorry, now I remember the problem! So we either have to provide two complete share functions (with quality selection; like "Share Video"/"play external" and "Share Url" or something) or use some kind of setting or toggle, which I would like to avoid.

**Comment by kernal64** [comment]
From the user's point of view, it makes no difference whether the video is to be shared for viewing or whether the URL is to be shared. I think two buttons would be confusing any may unnecessarily clutter the ui. Can't a unified share function offer both types off apps and detect what type of receiver the application is (player or downloader) and automatically take the correct branch?

By the way - I like the app very much. It has a clean interface and is very responsive.

Really cool is, when you share with the main share button in the actionbar to _pyLoad_, the movie name becomes automatically the packet name. When I only could coose the quality level... ;-)

**Comment by cemrich** [comment]
Okay, we have a dilemma here. From a technical point of view we cannot use a single share function without loosing apps.

* **Toolbar share icon**: Is using a very broad ACTION_SEND intent to share the video url. This intent can be handled by many apps (e.g. Chat Apps, Email, ...), but not all video players. My Galaxy S7 test device refuses to play back the video wiith this intent.
* **Share button with quality selection**: Is using the correct ACTION_VIEW intent with a "video/*" mime type. This should be supported by all video players and not many more apps.

Whenever I remove any of the share intents, some functionality is lost. There is no way I know of to combine these two Intents.

These are the options I see right now:
1. Using the toolbar share icon, always share the highest quality available. This is easy to implement and (I assume) expected behavior.
2. Using the toolbar share icon, show a quality selection dialog. I really, really would like to avoid this option.
3. Provide two different Intent options that reflect the technical differences, like "Share" and "Play externally". This would be a clean and "correct" solution but most likely confuse a lot of users.

**Comment by kernal64** [comment]
I'm actually not an Android-developer, but can't you combine the SEND and the VIEW intent as described on stack  overflow :

[How to combine ACTION_SEND and ACTION_VIEW intent types?](https://stackoverflow.com/questions/57500024/how-to-combine-action-send-and-action-view-intent-types)

`// Share
val sendIntent = Intent(Intent.ACTION_SEND)
sendIntent.setDataAndType(uri, MIME_PDF_TYPE)
sendIntent.putExtra(Intent.EXTRA_TEXT, "TEST")
sendIntent.putExtra(Intent.EXTRA_STREA, uri)
    
// Open
val openIntent = Intent(Intent.ACTION_VIEW)
openIntent.setDataAndType(uri, MIME_PDF_TYPE)
openIntent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
    
val chooserIntent = Intent.createChooser(sendIntent,activity.getString(R.string.sharing_title))
    
chooserIntent.putExtra(Intent.EXTRA_INITIAL_INTENTS, arrayOf(openIntent))
    
activity.startActivity(chooserIntent)`

**Comment by cemrich** [comment]
Wow, this was new to me. However, it does not work on Android 10 and up: https://stackoverflow.com/questions/59786777/why-intent-extra-initial-intents-not-working-in-android-10

**Comment by kernal64** [comment]
Oh, that's too bad that it no longer works on Android 10 and up. I installed Android Studio yesterday and am just getting to know it a bit. I've already cloned the repo and will have a look at it soon. Maybe I come up with another idea. If not, I think your first option is still the best. I would always prefer the highest quality, but I think that should be adjustable in the settings.
